crates/pallet-receipts/src/lib.rs

```diff
@@ -359,11 +359,12 @@ impl<T: Config> Pallet<T> {
             FraudProofError::ExecutionReceiptInFuture
         );
 
-        let parent_hash = T::Hash::decode(&mut fraud_proof.parent_hash.encode().as_slice())
-            .map_err(|_| FraudProofError::WrongHashType)?;
+        let primary_parent_hash =
+            T::Hash::decode(&mut fraud_proof.primary_parent_hash.encode().as_slice())
```

Comment:
Another lesson learned is that encoding/decoding erases types and can easily cause problems. We should consider adding to/from bounds and ensure that types are compatible statically even if we can't name them explicitly due to use of generics.

Reply:
Although I'm not super happy about adding the from/to bounds due to several reasons, I agree on the benefits gained from the compiler. We had a TODO, but haven't fixed it and applied the rule everythere.